pkg/resource/restconfig_factory.go

```diff
@@ -2,30 +2,71 @@ package resource
 
 import (
 	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 	"k8s.io/client-go/tools/clientcmd"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sync"
 )
 
+// HIVE-2399: In the cluster sync controller, getRESTConfigFactory is called once per
+// reconcile loop. Previously, a restConfigClientGetter was created for every reconcile,
+// used to apply (or rather, attempt to apply) various synced objects, and then thrown
+// away at the end of the reconcile pass.
+//
+// The restConfigClientGetter (or more accurately, the  cmdutil.Factory that wraps it) is
+// used by the cluster sync controller (via this helper) to apply the various objects to
+// the target cluster during the syncing process. During the application process (see apply.go),
+// the apply command is set up (aptly named setupApplyCommand()) using the factory returned
+// by the getRESTConfigFactory() below. Within this setup, the ToRESTMapper() function of the
+// restConfigClientGetter is invoked. While the mapper returned is a deferred variant,
+// eventually the inner NewDiscoveryRESTMapper() function is invoked, which is very expensive
+// and allocates a huge amount of memory over time if repeatedly called in a hot loop. A graphic
+// demonstrating these allocations is attached the the HIVE-2399 card.
+//
+// The solution is twofold:
+// 1) store the deferred mapper on the restConfigClientGetter so that it can be reused
+// 2) cache the restConfigClientGetter for each rest config and reuse it for future reconciles
+var restConfigClientGetterCache sync.Map
+
 func (r *helper) getRESTConfigFactory(namespace string) (cmdutil.Factory, error) {
 	if r.metricsEnabled {
 		// Copy the possibly shared restConfig reference and add a metrics wrapper.
 		cfg := rest.CopyConfig(r.restConfig)
 		controllerutils.AddControllerMetricsTransportWrapper(cfg, r.controllerName, false)
 		r.restConfig = cfg
 	}
-	r.logger.WithField("cache-dir", r.cacheDir).Debug("creating cmdutil.Factory from REST client config and cache directory")
-	f := cmdutil.NewFactory(&restConfigClientGetter{restConfig: r.restConfig, cacheDir: r.cacheDir, namespace: namespace})
+
+	// HIVE-2399: reuse the same restConfigClientGetter for each restConfig
+	restClientGetter, loaded := restConfigClientGetterCache.LoadOrStore(r.restConfig, &restConfigClientGetter{
```

Comment:
I'm pretty interested/concerned about how a rest.Config is used as a map key. Especially since some of the members are funcs. I guess as long as the funcs are static (not inlines/lambdas) they'll always hash the same? Except at least one of the funcs (cfg.Dial) looks like it's a lambda...

My point is: how do we know we're actually reusing entries?

Reply:
Long story short: it doesn't (and I want to discuss further)